src/main/java/org/elasticsearch/index/shard/ShardPath.java

```diff
@@ -0,0 +1,173 @@
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.index.shard;
+
+import org.elasticsearch.ElasticsearchIllegalStateException;
+import org.elasticsearch.cluster.metadata.IndexMetaData;
+import org.elasticsearch.common.collect.Tuple;
+import org.elasticsearch.common.logging.ESLogger;
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.env.NodeEnvironment;
+import org.elasticsearch.index.settings.IndexSettings;
+
+import java.io.IOException;
+import java.nio.file.FileStore;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+
+public final class ShardPath {
+    public static final String INDEX_FOLDER_NAME = "index";
+    public static final String TRANSLOG_FOLDER_NAME = "translog";
+
+    private final Path path;
+    private final String indexUUID;
+    private final ShardId shardId;
+    private final Path shardStatePath;
+
+
+    public ShardPath(Path path, Path shardStatePath, String indexUUID, ShardId shardId) {
+        this.path = path;
+        this.indexUUID = indexUUID;
+        this.shardId = shardId;
+        this.shardStatePath = shardStatePath;
+    }
+
+    public Path resolveTranslog() {
+        return path.resolve(TRANSLOG_FOLDER_NAME);
+    }
+
+    public Path resolveIndex() {
+        return path.resolve(INDEX_FOLDER_NAME);
+    }
+
+    public Path getDataPath() {
+        return path;
+    }
+
+    public boolean exists() {
+        return Files.exists(path);
+    }
+
+    public String getIndexUUID() {
+        return indexUUID;
+    }
+
+    public ShardId getShardId() {
+        return shardId;
+    }
+
+    public Path getShardStatePath() {
+        return shardStatePath;
+    }
+
+    /**
+     * This method walks through the nodes shard paths to find the data and state path for the given shard. If multiple
+     * directories with a valid shard state exist the one with the highest version will be used.
+     * <b>Note:</b> this method resolves custom data locations for the shard.
+     */
+    public static ShardPath loadShardPath(ESLogger logger, NodeEnvironment env, ShardId shardId, @IndexSettings Settings indexSettings) throws IOException {
+        final String indexUUID = indexSettings.get(IndexMetaData.SETTING_UUID, IndexMetaData.INDEX_UUID_NA_VALUE);
+        final Path[] paths = env.shardPaths(shardId);
+        Tuple<Path, ShardStateMetaData> metaDataTuple = null;
+        for (Path path : paths) {
+            ShardStateMetaData load = ShardStateMetaData.FORMAT.loadLatestState(logger, path);
+            if (load != null) {
+                if (metaDataTuple == null || metaDataTuple.v2().version < load.version) {
```

Comment:
multiple shard *states*. By this point the upgradeMultiDataPath code should have brought it down to a single shard state file, right?